### PR TITLE
Refactor plot utils

### PR DESCRIPTION
**Pull request by dcherian**
Making the Dataset plotting PR smaller. This one 
1. Refactors out plotting utility functions.
2. Refactors some facetgrid code.
3. Refactors out parsing colormap, colorbar kwargs
4. Refactors out making colorbars

### CODE REVIEW

**Comment by dcherian** [comment]
Tests should pass after #2673 is fixed.

File: xarray/plot/facetgrid.py [hunk 1]
```diff
@@ -221,95 +221,75 @@ def map_dataarray(self, func, x, y, **kwargs):
 
         """
 
-        cmapkw = kwargs.get('cmap')
-        colorskw = kwargs.get('colors')
-        cbar_kwargs = kwargs.pop('cbar_kwargs', {})
-        cbar_kwargs = {} if cbar_kwargs is None else dict(cbar_kwargs)
+        if func.__name__ == 'line':
```
[the remark this answers is not in the document]
Comment (dcherian) [reply xarray/plot/facetgrid.py]: OK I can undo that.

File: xarray/plot/facetgrid.py [hunk 2]
```diff
@@ -522,3 +489,33 @@ def map(self, func, *args, **kwargs):
...
+    if size is None:
+        size = 3
+    elif figsize is not None:
+        raise ValueError('cannot provide both `figsize` and `size` arguments')
+
+    g = FacetGrid(data=data, col=col, row=row, col_wrap=col_wrap,
+                  sharex=sharex, sharey=sharey, figsize=figsize,
+                  aspect=aspect, size=size, subplot_kws=subplot_kws)
+
+    if kind == 'line':
```
Comment (dcherian) [inline xarray/plot/facetgrid.py]: @shoyer do you prefer this way to specift which `map_` function to use?

Comment (dcherian) [reply xarray/plot/facetgrid.py]: ping @shoyer

Comment (shoyer) [reply xarray/plot/facetgrid.py]: Yes, this definitely looks cleaner to me

**Pull request merged**